feat(metrics): add language parse statistics

kopecs opened the pull request:
Closes PA-1678

PR checklist:

- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] `changelog.d/<issue>.<type>` is a file with the _what_, _why_, and _how_ of the change.
  - \<issue> is `pa-312` (Linear ticket), `gh-1234` (GitHub issue), or `new-gizmo` (unique semantic name)
  - \<type> is `added`, `changed`, `fixed`, or `infra`.
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Changelog guidelines](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry)
- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

---

This PR adds an additional piece of metics information---by language parse rates. This consists of a language aggregated object which reports the targeted files & byte counts, and the amount of files we were able to parse without error and the number of bytes parsed.

The purpose of this is to help drive parsing improvements in the future by better understanding which of our parsers have poor performance.

Commit:
Implement parse rate metrics

core: deduplicate partial parse errors which are currently duplicated on
a per-rule basis.

cli: adds ParsingData in a similar vein to the existing ProfilingData to
track information about parse rates by language to further help improve
our parsing capabilities by better understanding current parse rates.

Commit:
Update snapshots

Commit:
Update PRIVACY.md

Add entries and example for parsing statistics.

Commit:
Add changelog entry

kopecs commented:
Is the existing test coverage for metrics sufficient, or do we have tests on a per-metric basis?

Commit:
fixup! Implement parse rate metrics

emjin commented on semgrep-core/src/core/Report.ml:
```diff
@@ -208,7 +208,7 @@ let collate_results init_extra unzip_extra base_case_extra final_extra results =
   {
     matches = List.flatten matches;
-    errors = List.flatten errors;
+    errors = List.flatten errors |> List.sort_uniq compare;
```
Add a comment in `Semgrep_output_v0.atd` that we guarantee errors are sorted and deduped in core (though matches may not be), and then leave a corresponding comment here pointing users to the `Semgrep_output_v0.atd` comment

emjin replied:
I would also add in the comment here (`Report.ml`) that we dedup errors more aggressively to be consistent with parse errors

emjin replied:
Oh, one more thing---`collate_results` is run many times, since it's used by both `collate_rule_results` and `collate_pattern_results`. It's worth estimating the complexity of `List.sort_uniq` vs using a set

kopecs replied:
In terms of that last point, do the existing docs we have for perf cover what I would need to do to test that?

emjin replied:
They should; tldr is `./run-benchmarks` with `ci-large-repos.yaml`. If that doesn't make any sense even after reading docs, poke me to update them

emjin commented on semgrep-core/src/engine/Match_search_mode.ml:
```diff
@@ -134,7 +134,9 @@ let (group_matches_per_pattern_id : Pattern_match.t list -> id_to_match_results)
 let error_with_rule_id rule_id (error : E.error) =
-  { error with rule_id = Some rule_id }
+  match error.typ with
+  | PartialParsing _ -> error
```
Could use a comment that this is to be consistent with parse errors

emjin commented on cli/src/semgrep/parsing_data.py:
```diff
@@ -0,0 +1,88 @@
+            lang_parse_data.targets_with_errors += 1
+
+        # Total errors for the whole file
+        if isinstance(
```
I'd be more comfortable if this were encoded in the interface file. The situation I'm imagining is: say someone turns `ParseError` recoverable. How would they know to change this?

kopecs replied:
Encoded in the sense of just a comment on each of the variants in the atd file documenting what we use it for, or are you envisioning something else?

emjin replied:
I don't have an option I really like. At minimum a bidirectional comment. Another possibility is to include it as part of the error type, the way `core_severity` is. I think the lack of testing is part of what's making me skittish, so I'll give it some more thought

kopecs replied:
Including it with the error type feels more cumbersome to me, as it provides an additional field which has to get updated to be kept in a valid state (don't want to end up with `is_parse_error` as `true` but have the `error_kind` be `Timeout`).

Added a comment, and I'll increase test coverage.

emjin commented:
I don't know about metrics, but this does need a test to run semgrep on files with various kinds of errors (ideally exercise all five and ideally use files that are actually errors so that hopefully this test doesn't need to be changed later when an error disappears). It should enforce that the parse percentage reported is x%.

Commit:
Update submodule

Commit:
Add additional documentation about coupling

Because some metrics code now relies on the set of core error kinds
which comprise parse errors, as well as if a core error kind indicates a
total or partial parse-related error, we need additional documentation
to ensure this remains in sync.

Commit:
fixup! Implement parse rate metrics

Commit:
Add tests for parse metrics

Commit:
Use `Set` instead of `sort_uniq`

Benchmarks see slightly better performance on my machine for using a set
instead of `sort_uniq`.

With `sort_uniq`:
 - semgrep.bench.gitlab.std.duration = 2.628 s
 - semgrep.bench.njs-old-juice.std.duration = 1374.259 s

With `Set`:
 - semgrep.bench.gitlab.std.duration = 3.118 s
 - semgrep.bench.njs-old-juice.std.duration = 1309.339 s

kopecs commented:
Looks like CI failure is due to a lagging pr that got merged for semgrep-interfaces.

Commit:
Update semgrep_interfaces

Commit:
Merge branch 'develop' into cooper/parse-metrics
